File: Element/Climb_Stairs_k.cxx
/* Climb stairs with at most k steps. (k > 0)
 * Time complexity: O(n*k), where n is the number of staircases and k is the maximum step allowed.
*/

#include <bits/stdc++.h>
using namespace std;

class Solution{
	public:
	static int  climbStairs_k(int n, int k){
		vector<int> dp(n+1,0);
		dp[0] = 1;
		for(auto i(1); i <= n; ++i){
			for(auto j(1); j <= k; ++j){
				if (i-j >= 0){
					dp[i] += dp[i-j];
				}
			}
		}
		return dp[n];
	}
};

int main(){
	ios_base::sync_with_stdio(false);
	int n = 12;
	int k = 5;
	cout << Solution::climbStairs_k(n,k); // 1793
	return 0;
}

File: Element/Climb_Stairs_k.py
"""
Climb stairs with at most k steps. (k > 0)
Time complexity: O(n*k), where n is the number of staircases and k is the maximum step allowed.
"""

def climbStairs_k(n:int,k:int)->int:
	dp = [0 for x in range(n+1)]
	dp[0] = 1
	for i in range(1,n+1):
		for j in range(1,k+1):
			if i-j >= 0:
				dp[i] += dp[i-j]
	return dp[n]
	
if __name__=="__main__":
	n = 12
	k = 5
	print(climbStairs_k(n,k),end=" ") # 1793
